### PR TITLE
update ad-level so it uses a circular flowgraph

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Summary

This is a fix for the ad-level optimization. There was a bug where the monotone framework was not continuously iterating in the minimum fixed point algorithm to find the set of variables that can be made into data. This led to the algorithm stopping too early and possibly producing non-compilable C++ such as for the following model

```stan
data {
    matrix[10, 10] X_data;
}

parameters {
    matrix[10, 10] X_p;
}

transformed parameters {
    matrix[10, 10] X_tp1 = exp(X_data);
    matrix[10, 10] X_tp2 = exp(X_tp1);
    X_tp1 = X_p;
}
```

In the current impl the ad level optimization will think `X_tp1` is an autodiff type but `X_tp2` is data. Since we can't assign autodiff variables to data variables this would throw a C++ compiler error. 

By connecting the end nodes to the initial nodes the minimum fixed point algorithm will continously run until no changes to the program are made. Now the above program deduces that since `X_tp1` must be an autodiff type `X_tp2` must be as well.

In the test model you'll see that lazy code motion actually allows `X_tp2` to still be data, but that's because `exp(X_data)` is stored in a local variable and the code becomes the equivalent of 

```stan
matrix[10, 10] X_tp2 = exp(exp(X_data));
```

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

bug fix for ad-level optimization 

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by SteveBronder** [comment]
@WardBrian fyi the plan is to merge this, then `O1`, then the NA filling fix (minus turning it on by default and just the fix) then the `O1` PR

**Comment by SteveBronder** [comment]
@rybern fyi I set you as the reviewer since a lot of this is actually your code I just modified

**Comment by rybern** [comment]
Great, I'll have a look tomorrow

**Review by rybern — APPROVED** [review]
Looks good to me as long as those tests results are right, I didn't double check those.

One thing you could consider adding is an expect test showing that name collisions can happen with this approach. Maybe a program like

```
parameters {
  real x;
  real y;
}
model {
  if (x > 1) {
    int a = x;
    y ~ normal(a, 1);
  } else {
    int a = 1; // a is AD but doesn't need to be
    y ~ normal(a, 1);
  }
}
```
so we can see the progress when we/if we fix that

File: test/integration/good/compiler-optimizations/cpp.expected [hunk 1]
```diff
@@ -1,3 +1,1767 @@
...
+      Eigen::Matrix<double, -1, -1> lcm_sym10__;
+      Eigen::Matrix<double, -1, -1> lcm_sym9__;
+      Eigen::Matrix<local_scalar_t__, -1, -1> X_p;
+      current_statement__ = 1;
+      X_p = in__.template read<Eigen::Matrix<local_scalar_t__, -1, -1>>(10,
+              10);
+      Eigen::Matrix<local_scalar_t__, -1, -1> X_tp1;
+      current_statement__ = 2;
+      assign(X_tp1, X_d, "assigning variable X_tp1");
+      Eigen::Matrix<double, -1, -1> X_tp2;
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply test/integration/good/compiler-optimizations/cpp.expected:846]: X_tp1 is ad but the rest are not which, with lcm, is what we expect

[the remark this answers is not in the document]
Comment (SteveBronder) [reply test/integration/good/compiler-optimizations/cpp.expected:846]: X_tp1 is ad but the rest are not which, with lcm, is what we expect

**Comment by SteveBronder** [comment]
> One thing you could consider adding is an expect test showing that name collisions can happen with this approach. Maybe a program like

Let's leave that till @VMatthijs puts in that PR

File: src/analysis_and_optimization/Optimize.ml [hunk 2]
```diff
@@ -1024,54 +1024,98 @@ let block_fixing mir =
...
+       -> string Core_kernel.Set.Poly.t
+       -> ( Expr.Typed.Meta.t Expr.Fixed.t
+          , (Expr.Typed.Meta.t, 'a) Stmt.Fixed.t )
+          Stmt.Fixed.Pattern.t )
+    ~(extra_variables : string -> string Set.Poly.t)
+    ~(initial_variables : string Set.Poly.t) (stmt : Stmt.Located.t) =
+  let rev_flowgraph, flowgraph_to_mir =
+    Monotone_framework.inverse_flowgraph_of_stmt stmt in
+  let fwd_flowgraph = Monotone_framework.reverse rev_flowgraph in
+  let (module Fwd_Flowgraph) =
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply src/analysis_and_optimization/Optimize.ml:1061]: Word!

**Comment by SteveBronder** [comment]
Aight think we are good! @rybern Once test passes and you approve feel free to merge

**Pull request merged**